### PR TITLE
vllm internal prefix cache compatibility

### LINKED ISSUE #145: LMCache and VLLM's internal prefix caching need to be compatible

### PR DESCRIPTION
**Pull request by XbzOnGit**
Make LMCache work together with vllm internal prefix cache.  
This resolves LMCache/LMcache#145

### CODE REVIEW

File: lmcache_vllm/vllm_adapter.py [hunk 1]
```diff
@@ -310,68 +281,109 @@ def lmcache_retrieve_kv(
...
+            start_pos_list.append(start_pos)
+            skip_leading_tokens = seq_len - slen
+            assert skip_leading_tokens >= 0
+            if seq_group_metadata.is_prompt:
+                is_prefill.append(True)
+                logger.debug(f"skip_leading_tokens: {skip_leading_tokens}")
+                logger.debug(f"Tokens assigned to compute by vllm: {slen}")
+                logger.debug(f"full token length: {seq_len}")
+                # 1. Retrieve
+                kv_tuple, num_retrieved_tokens = engine.retrieve(full_token_tensor)
```
Comment (YaoJiayi) [inline lmcache_vllm/vllm_adapter.py]: We shouldn't retrieve the kv cache that's already in vllm. I added a mask to lmcache backend in another pr.
See https://github.com/LMCache/LMCache/pull/153 .

File: lmcache_vllm/vllm_adapter.py [hunk 2]
```diff
@@ -405,46 +421,54 @@ def build_partial_prefill_input(
...
         # Attn metadata-related
-        rebuilt_num_prefills += 1
-        rebuilt_num_prefill_tokens += q_len
-        new_slot_mapping = slot_mapping_flat[start_pos + num_computed_token : start_pos + num_token]
+        if is_prefill:
+            rebuilt_num_prefills += 1
+            rebuilt_num_prefill_tokens += q_len
+        else:
+            assert q_len == 1
+            rebuilt_num_decode_tokens += q_len
```
Comment (YaoJiayi) [inline lmcache_vllm/vllm_adapter.py]: why do we need rebuilt_num_decode_tokens?

File: lmcache_vllm/vllm_adapter.py [hunk 3]
```diff
@@ -100,21 +106,14 @@ def lmcache_should_retrieve(
...
-    return all([
-        is_prefill_run, not is_profile_run
-    ])
+    if is_profile_run:
+        return False
+    seq_group_list = model_input.seq_group_metadata_list
+    for seq_group in seq_group_list:
+        if seq_group.is_prompt:
+            return True
+    return False
```
Comment (YaoJiayi) [inline lmcache_vllm/vllm_adapter.py:116]: We are we removing the is_prefill_run check on the left?

File: lmcache_vllm/vllm_adapter.py [hunk 4]
```diff
@@ -213,59 +199,49 @@ def lmcache_store_kv(
...
-            kv_tuple_list.append(
-                    (key_cache[current_slot_mapping],
-                    value_cache[current_slot_mapping])
-                )
-
-    
-        engine.store(current_tokens.cpu(), tuple(kv_tuple_list), skip_existing = True, blocking = False)
+    seq_group_metadata_list = model_input.seq_group_metadata_list
+    for seq_group_metadata in seq_group_metadata_list:
+        for seqid, seq_data in seq_group_metadata.seq_data.items():
```
Comment (YaoJiayi) [inline lmcache_vllm/vllm_adapter.py:204]: Just curious. Why should we iterate over two loops?

**Review by YaoJiayi — COMMENTED** [review]
Great, thanks @XbzOnGit! I left some comments.

File: lmcache_vllm/vllm_adapter.py [hunk 5]
```diff
@@ -213,59 +199,49 @@ def lmcache_store_kv(
...
+
+
+                        kv_tuple_list.append(
+                                (key_cache[current_slot_mapping],
+                                value_cache[current_slot_mapping])
+                            )
+                    assert skip_leading_tokens == 0
+                    logger.debug(f"Store skips {skip_leading_tokens} and then stores {seq_len - skip_leading_tokens}")
+                    engine.store(current_tokens.cpu(), tuple(kv_tuple_list), 
+                                skip_existing = True, blocking = False)
```
Comment (YaoJiayi) [inline lmcache_vllm/vllm_adapter.py:244]: Just to make sure. Does "skip_leading_tokens" indicate we are not storing KV cache to lmcache if it's already in vllm?

Comment (XbzOnGit) [reply lmcache_vllm/vllm_adapter.py:244]: It means if it is already in LMCache, we can skip the step of copying it out of paged attention to reduce some memory copy

File: lmcache_vllm/vllm_adapter.py (hunk 4 above)
Comment (XbzOnGit) [reply lmcache_vllm/vllm_adapter.py:204]: One loop for sequence group, another loop for every sequence inside. I think one sequence group can have multiple sequences in decoding stage, according to its sampling methods?

File: lmcache_vllm/vllm_adapter.py (hunk 3 above)
Comment (XbzOnGit) [reply lmcache_vllm/vllm_adapter.py:116]: I think is_prompt itself should indicate if this sequence group is in prefill stage, according to the code in scheduler. So it should be the same with is_prefill_run.

File: lmcache_vllm/vllm_adapter.py (hunk 1 above)
Comment (XbzOnGit) [reply lmcache_vllm/vllm_adapter.py]: OK, I actually did that before. But to keep it compatible with the current LMCache, I removed it from the final PR.  

But one question, when do we need to mask out some tokens in the middle? When I implemented this, I just added an int number to skip some retrieving from the start.

File: lmcache_vllm/vllm_adapter.py (hunk 2 above)
Comment (XbzOnGit) [reply lmcache_vllm/vllm_adapter.py]: We don't need it. This can be removed.

File: lmcache_vllm/vllm_adapter.py (hunk 1 above)
Comment (YaoJiayi) [reply lmcache_vllm/vllm_adapter.py]: What's your take on this? I was thinking the more general case of non-prefix caching.

File: lmcache_vllm/vllm_adapter.py (hunk 5 above)
Comment (YaoJiayi) [reply lmcache_vllm/vllm_adapter.py:244]: You mean LMCache or vllm?

File: lmcache_vllm/vllm_adapter.py (hunk 4 above)
Comment (YaoJiayi) [reply lmcache_vllm/vllm_adapter.py:204]: Yes, I think it's for beam search. But isn't this gonna add a lot overhead from copying data from paged memory to a contiguous buffer?

File: lmcache_vllm/vllm_adapter.py (hunk 3 above)
Comment (YaoJiayi) [reply lmcache_vllm/vllm_adapter.py:116]: Ok. We can keep it this way for now. There should be a lot of conflicts between this and chunk prefill. But I can resolve the conflicts after merging this pr.

**Comment by YaoJiayi** [comment]
Also, could you add an offline example for this?

File: lmcache_vllm/vllm_adapter.py (hunk 5 above)
Comment (XbzOnGit) [reply lmcache_vllm/vllm_adapter.py:244]: I mean we should lookup LMCache first, for what tokens are already inside LMCache.  Then when we are building the kv_tuple, we can slice with a shorter slot_mapping(skipping those already in LMCache), then those pages will not even get sliced. But the current LMCache does not have lookup, so I just set it to zero.

File: lmcache_vllm/vllm_adapter.py (hunk 1 above)
Comment (XbzOnGit) [reply lmcache_vllm/vllm_adapter.py]: When I add this, I was thinking about efficiency. But according to your PR, the masking is just looking for the last token in a chunk and mask chunk by chunk. I guess the API is a little misleading. Like if the user sets the first 257 tokens to false, and chunk size is 256, it returns a tensor with a confusing shape.

**Review by YaoJiayi — APPROVED** [review]

**Pull request merged**